landlab/components/lake_fill/lake_fill_barnes.py

```diff
@@ -0,0 +1,2149 @@
+        are permitted.
+        """
+        return np.array(self._tasks_ever_in_queue)
+
+
+def _fill_one_node_to_flat(fill_surface, all_neighbors,
+                           pitq, openq, closedq, dummy):
+    """
+    Implements the Barnes et al. algorithm for a simple fill. Assumes the
+    _open and _closed lists have already been updated per Barnes algos 2&3,
+    lns 1-7.
+
+    Parameters
+    ----------
+    fill_surface : 1-D array of length nnodes
+        The surface to fill in LL node order. Modified in place.
+    all_neighbors : (nnodes, max_nneighbours) array
+        Adjacent nodes at each node.
+    pitq : heap queue (i.e., a structured list)
+        Current nodes known to be in a lake, if already identified.
+    openq : StablePriorityQueue object
+        Ordered queue of nodes remaining to be checked out by the algorithm
+        that are known not to be in a lake.
+    closedq : 1-D boolean array of length nnodes
+        Nodes already or not to be explored by the algorithm.
+    dummy : any Python object
+        Necessary for direct comparison with _fill_one_node_to_slant.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from landlab import RasterModelGrid, CLOSED_BOUNDARY
+    >>> mg = RasterModelGrid((5, 6), 1.)
+    >>> for edge in ('left', 'top', 'bottom'):
+    ...     mg.status_at_node[mg.nodes_at_edge(edge)] = CLOSED_BOUNDARY
+    >>> z = mg.zeros('node', dtype=float)
+    >>> z.reshape(mg.shape)[2, 1:-1] = [2., 1., 0.5, 1.5]
+    >>> z.reshape(mg.shape)[1, 1:-1] = [2.1, 1.1, 0.6, 1.6]
+    >>> z.reshape(mg.shape)[3, 1:-1] = [2.2, 1.2, 0.7, 1.7]
+    >>> zw = z.copy()
+    >>> openq = StablePriorityQueue()
+    >>> pitq = []
+    >>> closedq = mg.zeros('node', dtype=bool)
+    >>> closedq[mg.status_at_node == CLOSED_BOUNDARY] = True
+    >>> edges = np.array([11, 17, 23])
+    >>> for edgenode in edges:
+    ...     openq.add_task(edgenode, priority=z[edgenode])
+    >>> closedq[edges] = True
+    >>> while True:
+    ...     try:
+    ...         _fill_one_node_to_flat(zw, mg.adjacent_nodes_at_node,
+    ...                                pitq, openq, closedq, None)
+    ...     except KeyError:
+    ...         break
+
+    Now check the values make sense.
+
+    >>> lake = np.array([False, False, False, False, False, False,
+    ...                  False, False,  True,  True, False, False,
+    ...                  False, False,  True,  True, False, False,
+    ...                  False, False,  True,  True, False, False,
+    ...                  False, False, False, False, False, False])
+    >>> np.allclose(zw[lake], z[16])
+    True
+    >>> np.all(np.greater(zw[lake], z[lake]))
+    True
+    >>> np.allclose(zw[np.logical_not(lake)], z[np.logical_not(lake)])
+    True
+    """
+    try:
+        c = heapq.heappop(pitq)
+    except IndexError:
+        c = openq.pop_task()
+        # this will raise a KeyError once it's exhausted both queues
```

Comment:
Think I prefer this way, given the naming of the function. One call should do one thing, right? The loop in the example is explicitly raising each cell, one after the other, so the while look sits there. I guess it could return a break, but I find that a little less intuitive.